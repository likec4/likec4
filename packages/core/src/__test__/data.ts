import { indexBy } from 'remeda'
import { ModelIndex } from '../model-index'
import type { Element, ElementKind, Fqn, Relation, RelationID, Tag } from '../types'

/**
              ┌──────────────────────────────────────────────────┐
              │                      cloud                       │
              │  ┌───────────────────────────────────────────┐   │
              │  │                 frontend                  │   │
┏━━━━━━━━━━┓  │  │   ┏━━━━━━━━━━━━━┓   ┏━━━━━━━━━━━━━━━━┓    │   │   ┏━━━━━━━━━━━┓
┃          ┃  │  │   ┃             ┃   ┃                ┃    │   │   ┃           ┃
┃ customer ┃──┼──┼──▶┃  dashboard  ┃   ┃   adminpanel   ┃◀───┼───┼───┃  support  ┃
┃          ┃  │  │   ┃             ┃   ┃                ┃    │   │   ┃           ┃
┗━━━━━━━━━━┛  │  │   ┗━━━━━━┳━━━━━━┛   ┗━━━━━━━━┳━━━━━━━┛    │   │   ┗━━━━━━━━━━━┛
              │  └──────────┼───────────────────┼────────────┘   │
              │             ├───────────────────┘                │
              │             │                                    │
              │  ┌──────────┼────────────────────────────────┐   │
              │  │          ▼       backend                  │   │
              │  │   ┏━━━━━━━━━━━━━┓       ┏━━━━━━━━━━━━━┓   │   │
              │  │   ┃             ┃       ┃             ┃   │   │
              │  │   ┃  graphlql   ┃──────▶┃   storage   ┃   │   │
              │  │   ┃             ┃       ┃             ┃   │   │
              │  │   ┗━━━━━━━━━━━━━┛       ┗━━━━━━┳━━━━━━┛   │   │
              │  └────────────────────────────────┼──────────┘   │
              └───────────────────────────────────┼──────────────┘
                                                  │
                                        ┌─────────┼─────────┐
                                        │ amazon  │         │
                                        │         ▼         │
                                        │ ┏━━━━━━━━━━━━━━┓  │
                                        │ ┃              ┃  │
                                        │ ┃      s3      ┃  │
                                        │ ┃              ┃  │
                                        │ ┗━━━━━━━━━━━━━━┛  │
                                        └───────────────────┘

specification {
  element actor
  element system
  element container
  element component

  tag old
}

model {

  actor customer
  actor support

  system cloud {
    container backend {
      component graphql
      component storage {
        #old
      }

      graphql -> storage
    }

    container frontend {
      component dashboard {
        -> graphql
      }
      component adminPanel {
        #old
        -> graphql
      }
    }
  }

  customer -> dashboard
  support -> adminPanel

  system amazon {
    component s3

    cloud.backend.storage -> s3
  }

}

 */
const el = ({
  id,
  kind,
  title,
  ...props
}: Partial<Omit<Element, 'id' | 'kind'>> & { id: string; kind: string }): Element => ({
  id: id as Fqn,
  kind: kind as ElementKind,
  title: title ?? id,
  description: null,
  technology: null,
  tags: null,
  links: null,
  ...props
})

export const fakeElements = {
  amazon: el({
    id: 'amazon',
    kind: 'system',
    title: 'amazon'
  }),
  cloud: el({
    id: 'cloud',
    kind: 'system',
    title: 'cloud'
  }),
  customer: el({
    id: 'customer',
    kind: 'actor',
    title: 'customer',
    shape: 'person'
  }),
  support: el({
    id: 'support',
    kind: 'actor',
    title: 'support',
    shape: 'person'
  }),
  'amazon.s3': el({
    id: 'amazon.s3',
    kind: 'component',
    title: 's3',
    shape: 'storage'
  }),
  'cloud.backend': el({
    id: 'cloud.backend',
    kind: 'container',
    title: 'backend'
  }),
  'cloud.frontend': el({
    id: 'cloud.frontend',
    kind: 'container',
    title: 'frontend',
    shape: 'browser'
  }),
  'cloud.backend.graphql': el({
    id: 'cloud.backend.graphql',
    kind: 'component',
    title: 'graphql'
  }),
  'cloud.backend.storage': el({
    id: 'cloud.backend.storage',
    kind: 'component',
    title: 'storage',
    tags: ['old' as Tag]
  }),
  'cloud.frontend.adminPanel': el({
    id: 'cloud.frontend.adminPanel',
    kind: 'component',
    title: 'adminPanel',
    tags: ['old' as Tag]
  }),
  'cloud.frontend.dashboard': el({
    id: 'cloud.frontend.dashboard',
    kind: 'component',
    title: 'dashboard'
  })
} satisfies Record<string, Element>

export type FakeElementIds = keyof typeof fakeElements

const rel = ({
  source,
  target,
  title
}: {
  source: FakeElementIds
  target: FakeElementIds
  title?: string
}): Relation => ({
  id: `${source}:${target}` as RelationID,
  title: title ?? '',
  source: source as Fqn,
  target: target as Fqn
})

export const fakeRelations = [
  rel({
    source: 'customer',
    target: 'cloud.frontend.dashboard'
  }),
  rel({
    source: 'support',
    target: 'cloud.frontend.adminPanel'
  }),
  rel({
    source: 'cloud.backend.storage',
    target: 'amazon.s3'
  }),
  rel({
    source: 'cloud.backend.graphql',
    target: 'cloud.backend.storage'
  }),
  rel({
    source: 'cloud.frontend.dashboard',
    target: 'cloud.backend.graphql'
  }),
  rel({
    source: 'cloud.frontend.adminPanel',
    target: 'cloud.backend.graphql'
  })
]

export type FakeRelationIds = keyof typeof fakeRelations

export const fakeModel = () =>
  ModelIndex.from({
    elements: fakeElements,
    relations: indexBy(fakeRelations, r => r.id)
  })
