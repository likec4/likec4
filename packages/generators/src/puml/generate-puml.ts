import type {
  ComputedEdge,
  ComputedNode,
  ComputedView,
  ElementThemeColorValues,
  KeysOf,
  NodeId,
  RelationshipThemeColorValues,
  ThemeColorValues,
} from '@likec4/core'
import { CompositeGeneratorNode, joinToNode, NL, toString } from 'langium/generate'
import { isNullish as isNil } from 'remeda'

const capitalizeFirstLetter = (value: string) => value.charAt(0).toLocaleUpperCase() + value.slice(1)

const fqnName = (nodeId: string): string => nodeId.split('.').map(capitalizeFirstLetter).join('')

const nodeName = (node: ComputedNode): string => {
  return fqnName(node.parent ? node.id.slice(node.parent.length + 1) : node.id)
}

const pumlColor = (
  color: string | undefined,
  customColorProvider: (colorKey: string) => string | undefined,
  defaultColor: string = '#3b82f6',
) => {
  switch (color) {
    case 'blue':
    case 'primary': {
      return '#3b82f6'
    }
    case 'amber': {
      return '#a35829'
    }
    case 'gray': {
      return '#737373'
    }
    case 'green': {
      return '#428a4f'
    }
    case 'indigo': {
      return '#6366f1'
    }
    case 'slate':
    case 'muted': {
      return '#64748b'
    }
    case 'red': {
      return '#ac4d39'
    }
    case 'sky':
    case 'secondary': {
      return '#0284c7'
    }
    case null:
    case undefined: {
      return defaultColor
    }
    default:
      return customColorProvider(color) || color
  }
}

const pumlDirection = ({ autoLayout }: ComputedView) => {
  switch (autoLayout.direction) {
    case 'TB': {
      return 'top to bottom'
    }
    case 'BT': {
      console.warn('Bottom to top direction is not supported. Defaulting to top to bottom.')
      return 'top to bottom'
    }
    case 'LR': {
      return 'left to right'
    }
    case 'RL': {
      console.warn('Right to left direction is not supported. Defaulting to left to right.')
      return 'left to right'
    }
  }
}

const pumlShape = ({ shape }: ComputedNode) => {
  switch (shape) {
    case 'queue':
    case 'rectangle':
    case 'person': {
      return shape
    }
    case 'storage':
    case 'cylinder': {
      return 'database' as const
    }
    case 'mobile':
    case 'browser': {
      return 'rectangle' as const
    }
  }
}

const escapeLabel = (label: string | null | undefined) => isNil(label) ? null : JSON.stringify(label).slice(1, -1)

export function generatePuml<V extends ComputedView>(view: V) {
  const { nodes, edges, customColorDefinitions } = view
  const customColors = new Map<string, ThemeColorValues>(Object.entries(customColorDefinitions || {}))
  const elemntColorProvider = (key: KeysOf<ElementThemeColorValues>) => (colorKey: string) =>
    customColors.get(colorKey)?.elements[key]
  const relationshipsColorProvider = (key: KeysOf<RelationshipThemeColorValues>) => (colorKey: string) =>
    customColors.get(colorKey)?.relationships[key]
  const names = new Map<NodeId, string>()

  const printHeader = () => {
    return new CompositeGeneratorNode()
      .append('title "', view.title || view.id, '"', NL)
      .append(pumlDirection(view), ' direction', NL)
  }

  const printTheme = () => {
    return new CompositeGeneratorNode()
      .append('hide stereotype', NL)
      .append('skinparam ranksep ', '60', NL)
      .append('skinparam nodesep ', '30', NL)
      .append('skinparam {', NL)
      .indent({
        indentedChildren: indent =>
          indent
            .append('arrowFontSize ', '10', NL)
            .append('defaultTextAlignment ', 'center', NL)
            .append('wrapWidth ', '200', NL)
            .append('maxMessageSize ', '100', NL)
            .append('shadowing ', 'false', NL),
        indentation: 2,
      })
      .append('}', NL)
  }

  const printStereotypes = (node: ComputedNode): CompositeGeneratorNode => {
    const shape = pumlShape(node)
    const fqn = fqnName(node.id)

    return new CompositeGeneratorNode()
      .append('skinparam ', shape, '<<', fqn, '>>', '{', NL)
      .indent({
        indentedChildren: indent =>
          indent
            .append('BackgroundColor ', pumlColor(node.color, elemntColorProvider('fill')), NL)
            .append(
              'FontColor ',
              customColors.has(node.color)
                ? pumlColor(node.color, elemntColorProvider('hiContrast'))
                : '#FFFFFF',
              NL,
            )
            .append('BorderColor ', pumlColor(node.color, elemntColorProvider('stroke')), NL),
        indentation: 2,
      })
      .append('}', NL)
  }

  const printNode = (node: ComputedNode): CompositeGeneratorNode => {
    const shape = pumlShape(node)
    const fqn = fqnName(node.id)
    const label = escapeLabel(node.title) || nodeName(node)
    const tech = escapeLabel(node.technology)
    names.set(node.id, fqn)

    return new CompositeGeneratorNode()
      .append(shape, ' ')
      .append('"')
      .append('==', label)
      .appendIf(!!tech, '\\n', '<size:10>[', tech!, ']</size>')
      .appendIf(!!node.description, '\\n\\n', escapeLabel(node.description)!)
      .append('"', ' <<', fqn, '>> ', 'as ', fqn, NL)
  }

  const printBoundary = (node: ComputedNode): CompositeGeneratorNode => {
    const label = escapeLabel(node.title) || nodeName(node)
    const fqn = fqnName(node.id)
    names.set(node.id, fqn)

    return new CompositeGeneratorNode()
      .append('rectangle "', label, '" <<', fqn, '>> as ', fqn, ' {', NL)
      .indent({
        indentedChildren: indent =>
          indent
            .append(
              'skinparam ',
              'RectangleBorderColor<<',
              fqn,
              '>> ',
              pumlColor(node.color, elemntColorProvider('fill')),
              NL,
            )
            .append(
              'skinparam ',
              'RectangleFontColor<<',
              fqn,
              '>> ',
              pumlColor(node.color, elemntColorProvider('fill')),
              NL,
            )
            .append('skinparam ', 'RectangleBorderStyle<<', fqn, '>> ', 'dashed', NL, NL)
            .append(joinToNode(
              nodes.filter(n => n.parent === node.id),
              c => c.children.length > 0 ? printBoundary(c) : printNode(c),
            )),
        indentation: 2,
      })
      .append('}', NL)
  }

  const printEdge = (edge: ComputedEdge): CompositeGeneratorNode => {
    const tech = escapeLabel(edge.technology) || ''
    const label = escapeLabel(edge.label) || ''
    const color = pumlColor(edge.color, relationshipsColorProvider('lineColor'), '#777777')

    const colorTag = (color: string) => `<color:${color}>`

    return new CompositeGeneratorNode()
      .append(names.get(edge.source), ' .[', color, ',thickness=2].> ', names.get(edge.target))
      .appendIf(!!(label || tech), ' : "', colorTag(color))
      .appendIf(!!label, label, colorTag(color))
      .appendIf(!!(label && tech), '\\n')
      .appendIf(!!tech, colorTag(color), '<size:8>[', tech, ']</size>')
      .appendIf(!!(label || tech), '"')
      .append(NL)
  }

  return toString(
    new CompositeGeneratorNode()
      .append('@startuml', NL)
      .append(printHeader(), NL)
      .append(printTheme(), NL)
      .append(
        joinToNode(
          nodes.filter(n => n.children.length == 0),
          n => printStereotypes(n),
          {
            appendNewLineIfNotEmpty: true,
          },
        ),
      )
      .append(
        joinToNode(
          nodes.filter(n => isNil(n.parent)),
          n => n.children.length > 0 ? printBoundary(n) : printNode(n),
          {
            appendNewLineIfNotEmpty: true,
          },
        ),
      )
      .appendIf(
        edges.length > 0,
        NL,
        joinToNode(edges, e => printEdge(e), {
          appendNewLineIfNotEmpty: true,
        }),
      )
      .append(`@enduml`, NL),
  )
}
