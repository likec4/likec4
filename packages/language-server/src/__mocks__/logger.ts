import { vi } from 'vitest'
import type { Logger } from '../logger'

export const logger = {
  debug: vi.fn(),
  info: vi.fn(),
  warn: vi.fn(),
  log: vi.fn(),
  error: vi.fn(),
  trace: vi.fn()
} satisfies Logger
