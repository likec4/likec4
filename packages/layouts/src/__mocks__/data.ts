import { computeView, assignNavigateTo } from '@likec4/core'
import { cloudView, cloud3levels, amazonView, fakeModel, indexView } from './model'

const model = fakeModel()

export const [
  computedIndexView,
  computedCloudView,
  computedCloud3levels,
  computedAmazonView
] = assignNavigateTo([
  computeView(indexView, model)!,
  computeView(cloudView, model)!,
  computeView(cloud3levels, model)!,
  computeView(amazonView, model)!
])
