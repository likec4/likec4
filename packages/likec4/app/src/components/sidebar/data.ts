import type { DiagramView } from '@likec4/core'
import type { TreeNodeData } from '@mantine/core'
import { useMemo } from 'react'
import { find, isTruthy, values } from 'remeda'
import { useLikeC4Views } from 'virtual:likec4/store'

interface DiagramTreeNodeData {
  label: string
  value: string
  type: 'file' | 'folder' | 'diagram'
  children: DiagramTreeNodeData[]
}

export const isTreeNodeData = (node: TreeNodeData): node is DiagramTreeNodeData =>
  'type' in node && ['file', 'folder', 'diagram'].includes(node.type as any)

function compareTreeNodes(a: DiagramTreeNodeData, b: DiagramTreeNodeData) {
  if (a.children.length === 0 && b.children.length > 0) {
    return 1
  }
  if (a.children.length > 0 && b.children.length === 0) {
    return -1
  }
  return a.label.localeCompare(b.label)
}

function buildDiagramTreeData(views: DiagramView[]): DiagramTreeNodeData[] {
  const root: DiagramTreeNodeData = {
    value: '',
    label: 'Diagrams',
    type: 'folder',
    children: []
  }

  const findParent = (path: string): DiagramTreeNodeData => {
    let parent = root
    if (path === '') {
      return parent
    }
    const segments = path.split('/')
    const traversed = ['@fs'] as string[]
    while (segments.length) {
      const label = segments.shift() as string
      traversed.push(label)
      const value = traversed.join('/')
      let node = find(parent.children!, n => n.value === value)
      if (!node) {
        node = { label, value, type: 'folder', children: [] }
        parent.children.push(node)
      }
      parent = node
    }
    return parent
  }

  for (const view of views) {
    const parent = findParent(view.relativePath ?? '')
    parent.children.push({
      value: view.id,
      label: view.title ?? view.id,
      type: 'diagram',
      children: []
    })
    if (parent !== root) {
      parent.children.sort(compareTreeNodes)
      if (isTruthy(view.relativePath) && parent.type !== 'file') {
        parent.type = 'file'
      }
    }
  }

  return root.children.sort(compareTreeNodes)
}

// const $diagramsTree = batched($views, views => buildDiagramTreeData(values(views)))

export function useDiagramsTreeData() {
  const views = useLikeC4Views() as unknown as Record<string, DiagramView>
  return useMemo(() => buildDiagramTreeData(values(views)), [views])
}
