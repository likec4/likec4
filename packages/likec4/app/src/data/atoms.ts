import type { ExtractAtomValue } from 'jotai'
import { atom } from 'jotai'
import { atomFamily, atomWithReducer, splitAtom } from 'jotai/utils'
import { equals, groupBy, mapObject, values } from 'rambdax'
import { LikeC4Views } from '~likec4'
import { buildDiagramTreeAtom } from './sidebar-diagram-tree'
import type { DiagramView } from '@likec4/core'

function atomWithCompare<Value>(initialValue: Value) {
  return atomWithReducer(initialValue, (prev: Value, next: Value) => {
    if (equals(prev, next)) {
      return prev
    }
    return next
  })
}

const likec4ViewFamily = atomFamily(
  (view: DiagramView) => {
    const atom = atomWithCompare(view)
    atom.debugLabel = `view#${view.id}`
    return atom
  },
  (a, b) => a.id === b.id
)

const mapToAtoms = (views: typeof LikeC4Views) => mapObject(view => likec4ViewFamily(view), views)

const _viewsAtom = atom(mapToAtoms(LikeC4Views))
_viewsAtom.debugLabel = '_views'

export const viewsAtom = atom(
  get => get(_viewsAtom),
  (_, set, update: typeof LikeC4Views) => {
    set(
      _viewsAtom,
      mapObject(view => {
        const atom = likec4ViewFamily(view)
        set(atom, view)
        return atom
      }, update)
    )
  }
)
viewsAtom.debugLabel = 'views'

const indexPageTilesAtom = atom(get => {
  const views = values(get(viewsAtom))
  const byPath = groupBy(v => get(v).relativePath ?? '', views)
  return Object.entries(byPath)
    .map(([path, views]) => ({
      path,
      isRoot: path === '',
      views
    }))
    .sort((a, b) => {
      return a.path.localeCompare(b.path)
    })
})
export type IndexPageTile = ExtractAtomValue<typeof indexPageTilesAtom>[number]

const byPath = (tile: IndexPageTile) => tile.path
export const indexPageTilesAtomsAtom = splitAtom(indexPageTilesAtom, byPath)

// export type DashboardTile = ExtractAtomValue<typeof dashboardTilesAtom>[number]

// const key = (tile: DashboardTile) => tile.path

// export const diagramsTreeAtom = atom(buildDiagramTreeAtom(LikeC4Views))
export const diagramsTreeAtom = atom(get => {
  const views = values(get(viewsAtom))
  return buildDiagramTreeAtom(views.map(v => get(v)))
})

export const selectLikeC4ViewAtom = (viewId: string) => {
  return atom(get => {
    const viewAtom = get(viewsAtom)[viewId]
    return viewAtom ? get(viewAtom) : null
  })
}

if (import.meta.hot) {
  let $updateViews: undefined | ((update: typeof LikeC4Views) => void)

  viewsAtom.onMount = set => {
    $updateViews = set
    return () => {
      $updateViews = undefined
    }
  }

  import.meta.hot.accept('/@vite-plugin-likec4/likec4-generated', md => {
    const update = md?.LikeC4Views as typeof LikeC4Views | undefined
    if (update) {
      $updateViews?.(update)
    }
  })
}
