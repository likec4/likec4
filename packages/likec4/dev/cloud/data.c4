model {

  extend cloud.data {

    hitsAnomalies = app 'Hits Anomalies' {
      description '
        Process page hits.
        Search for anomalies.
      '

      -> amazon.sqs.hits 'process hits data'
    }

  }

}

views {

  view cloud_data of cloud.data {
    title "[context] Data Services"
    include
        *,
        cloud,
        amazon,
        amazon.sqs,
        cloud.data -> amazon.sqs.*

    style cloud {
      color muted
    }
    style amazon.* {
      color secondary
    }
  }

}
