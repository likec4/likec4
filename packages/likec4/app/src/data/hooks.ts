import { useAtomValue } from 'jotai'
import { useMemo } from 'react'
import { diagramsTreeAtom, indexPageTilesAtomsAtom, selectLikeC4ViewAtom } from './atoms'

export const useLikeC4View = (viewId: string) => {
  const anAtom = useMemo(() => selectLikeC4ViewAtom(viewId), [viewId])
  return useAtomValue(anAtom)
}

export const useIndexPageTileAtoms = () => {
  return useAtomValue(indexPageTilesAtomsAtom)
}

export type { IndexPageTile } from './atoms'

export const useDiagramsTree = () => useAtomValue(diagramsTreeAtom)
