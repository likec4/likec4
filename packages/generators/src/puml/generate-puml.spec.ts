import { expect, test } from 'vitest'
import { fakeComputedView3Levels, fakeDiagram, fakeDiagram2 } from '../__mocks__/data'
import { generatePuml } from './generate-puml'

test('generate puml - fakeDiagram', () => {
  expect(generatePuml(fakeDiagram)).toMatchSnapshot()
})

test('generate puml - fakeDiagram2', () => {
  expect(generatePuml(fakeDiagram2)).toMatchSnapshot()
})

test('generate puml - fakeComputedView 3 Levels', () => {
  expect(generatePuml(fakeComputedView3Levels)).toMatchSnapshot()
})
