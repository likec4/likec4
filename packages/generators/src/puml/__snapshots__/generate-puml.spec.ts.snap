// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generate puml - fakeComputedView 3 Levels 1`] = `
"@startuml
title "Context: Cloud"
top to bottom direction

hide stereotype
skinparam ranksep 60
skinparam nodesep 30
skinparam {
  arrowFontSize 10
  defaultTextAlignment center
  wrapWidth 200
  maxMessageSize 100
  shadowing false
}

skinparam rectangle<<Amazon>>{
  BackgroundColor #3b82f6
  FontColor #FFFFFF
  BorderColor #3b82f6
}
skinparam rectangle<<Customer>>{
  BackgroundColor #3b82f6
  FontColor #FFFFFF
  BorderColor #3b82f6
}
skinparam rectangle<<Support>>{
  BackgroundColor #3b82f6
  FontColor #FFFFFF
  BorderColor #3b82f6
}
skinparam rectangle<<CloudBackendGraphql>>{
  BackgroundColor #3b82f6
  FontColor #FFFFFF
  BorderColor #3b82f6
}
skinparam rectangle<<CloudBackendStorage>>{
  BackgroundColor #3b82f6
  FontColor #FFFFFF
  BorderColor #3b82f6
}
skinparam rectangle<<CloudFrontendAdminPanel>>{
  BackgroundColor #3b82f6
  FontColor #FFFFFF
  BorderColor #3b82f6
}
skinparam rectangle<<CloudFrontendDashboard>>{
  BackgroundColor #3b82f6
  FontColor #FFFFFF
  BorderColor #3b82f6
}
rectangle "==amazon" <<Amazon>> as Amazon
rectangle "cloud" <<Cloud>> as Cloud {
  skinparam RectangleBorderColor<<Cloud>> #3b82f6
  skinparam RectangleFontColor<<Cloud>> #3b82f6
  skinparam RectangleBorderStyle<<Cloud>> dashed

  rectangle "backend" <<CloudBackend>> as CloudBackend {
    skinparam RectangleBorderColor<<CloudBackend>> #3b82f6
    skinparam RectangleFontColor<<CloudBackend>> #3b82f6
    skinparam RectangleBorderStyle<<CloudBackend>> dashed

    rectangle "==graphql" <<CloudBackendGraphql>> as CloudBackendGraphql
    rectangle "==storage" <<CloudBackendStorage>> as CloudBackendStorage
  }
  rectangle "==adminPanel" <<CloudFrontendAdminPanel>> as CloudFrontendAdminPanel
  rectangle "==dashboard" <<CloudFrontendDashboard>> as CloudFrontendDashboard
}
rectangle "==customer" <<Customer>> as Customer
rectangle "==support" <<Support>> as Support

CloudFrontendDashboard .[#777777,thickness=2].> CloudBackendGraphql
CloudFrontendAdminPanel .[#777777,thickness=2].> CloudBackendGraphql
CloudBackendStorage .[#777777,thickness=2].> Amazon
CloudBackendGraphql .[#777777,thickness=2].> CloudBackendStorage
Support .[#777777,thickness=2].> CloudFrontendAdminPanel
Customer .[#777777,thickness=2].> CloudFrontendDashboard
@enduml
"
`;

exports[`generate puml - fakeDiagram 1`] = `
"@startuml
title "fakeView"
top to bottom direction

hide stereotype
skinparam ranksep 60
skinparam nodesep 30
skinparam {
  arrowFontSize 10
  defaultTextAlignment center
  wrapWidth 200
  maxMessageSize 100
  shadowing false
}

skinparam rectangle<<Amazon>>{
  BackgroundColor #3b82f6
  FontColor #FFFFFF
  BorderColor #3b82f6
}
skinparam rectangle<<Customer>>{
  BackgroundColor #3b82f6
  FontColor #FFFFFF
  BorderColor #3b82f6
}
skinparam rectangle<<Support>>{
  BackgroundColor #3b82f6
  FontColor #FFFFFF
  BorderColor #3b82f6
}
skinparam rectangle<<CloudBackend>>{
  BackgroundColor #3b82f6
  FontColor #FFFFFF
  BorderColor #3b82f6
}
skinparam rectangle<<CloudFrontend>>{
  BackgroundColor #3b82f6
  FontColor #FFFFFF
  BorderColor #3b82f6
}
rectangle "==amazon" <<Amazon>> as Amazon
rectangle "cloud" <<Cloud>> as Cloud {
  skinparam RectangleBorderColor<<Cloud>> #3b82f6
  skinparam RectangleFontColor<<Cloud>> #3b82f6
  skinparam RectangleBorderStyle<<Cloud>> dashed

  rectangle "==backend" <<CloudBackend>> as CloudBackend
  rectangle "==frontend" <<CloudFrontend>> as CloudFrontend
}
rectangle "==customer" <<Customer>> as Customer
rectangle "==support" <<Support>> as Support

CloudFrontend .[#777777,thickness=2].> CloudBackend : "<color:#777777>requests<color:#777777>"
CloudBackend .[#777777,thickness=2].> Amazon
Support .[#777777,thickness=2].> CloudFrontend
Customer .[#777777,thickness=2].> CloudFrontend : "<color:#777777>opens<color:#777777>"
@enduml
"
`;

exports[`generate puml - fakeDiagram2 1`] = `
"@startuml
title "frontend"
top to bottom direction

hide stereotype
skinparam ranksep 60
skinparam nodesep 30
skinparam {
  arrowFontSize 10
  defaultTextAlignment center
  wrapWidth 200
  maxMessageSize 100
  shadowing false
}

skinparam rectangle<<Client>>{
  BackgroundColor #3b82f6
  FontColor #FFFFFF
  BorderColor #3b82f6
}
skinparam rectangle<<SystemBackend>>{
  BackgroundColor #3b82f6
  FontColor #FFFFFF
  BorderColor #3b82f6
}
skinparam rectangle<<SystemFrontend>>{
  BackgroundColor #3b82f6
  FontColor #FFFFFF
  BorderColor #3b82f6
}
rectangle "==client" <<Client>> as Client
rectangle "==backend" <<SystemBackend>> as SystemBackend
rectangle "==frontend" <<SystemFrontend>> as SystemFrontend

SystemFrontend .[#777777,thickness=2].> SystemBackend : "<color:#777777>requests<color:#777777>"
Client .[#777777,thickness=2].> SystemFrontend : "<color:#777777>opens<color:#777777>"
@enduml
"
`;
